Device Plugin Usage documentation from GPU Operator (#101)
ROCm/gpu-operator: docs/device_plugin/device-plugin.md

| **EnableNodeLabeller**           | Enable/Disable node labeller with True/False |
| **DevicePluginArguments**        | The flag/values to pass on to Device Plugin  |
</br>

1. Both the `ImagePullPolicy` fields default to `Always` if `:latest` tag is specified on the respective Image, or defaults to `IfNotPresent` otherwise. This is default k8s behaviour for `ImagePullPolicy`

2. `DevicePluginArguments` is of type `map[string]string`. Currently supported key value pairs to set under `DevicePluginArguments` are:
   -> "resource_naming_strategy": {"single", "mixed"}

## How to choose Resource Naming Strategy

To customize the way device plugin reports gpu resources to kubernetes as allocatable k8s resources, use the `single` or `mixed` resource naming strategy in **DeviceConfig** CR
Before understanding each strategy, please note the definition of homogeneous and heterogeneous nodes

Homogeneous node: A node whose gpu's follow the same compute-memory partition style 
    -> Example: A node of 8 GPU's where all 8 GPU's are following CPX-NPS4 partition style
    
Heterogeneous node: A node whose gpu's follow different compute-memory partition styles

>     -> Example: A node of 8 GPU's where 5 GPU's are following SPX-NPS1 and 3 GPU's are following CPX-NPS1

Reviewer: Can we update this to say CPX-NPS4 as that will be the supported configuration.

```suggestion
    -> Example: A node of 8 GPU's where 5 GPU's are following SPX-NPS1 and 3 GPU's are following CPX-NPS4
```

Author: Hey Farshad, I wanted to give an example of heterogeneous. If we use cpx nps4, automatically everything moves to cps. The only way to have spx and cpx together is with nps 1, hence the use of nps to show heterogeneous examples

### Single

In `single` mode, the device plugin reports all gpu's (regardless of whether they are whole gpu's or partitions of a gpu) under the resource name `amd.com/gpu`
This mode is supported for homogeneous nodes but not supported for heterogeneous nodes

A node which has 8 GPUs where all GPUs are not partitioned will report its resources as:

```bash
amd.com/gpu: 8
```

A node which has 8 GPUs where all GPUs are partitioned using CPX-NPS4 style will report its resources as:

```bash
amd.com/gpu: 64
```

### Mixed

In `mixed` mode, the device plugin reports all gpu's under a name which matches its partition style.
This mode is supported for both homogeneous nodes and heterogeneous nodes

A node which has 8 GPUs which are all partitioned using CPX-NPS4 style will report its resources as:

```bash
amd.com/cpx_nps4: 64
```

> A node which has 8 GPUs where 5 GPU's are following SPX-NPS1 and 3 GPU's are following CPX-NPS1 will report its resources as:
> 
> ```bash
> amd.com/spx_nps1: 5
> amd.com/cpx_nps1: 24
> ``` 

Reviewer: Same here, please use CPX-NPS4

```suggestion
A node which has 8 GPUs where 5 GPU's are following SPX-NPS1 and 3 GPU's are following CPX-NPS4 will report its resources as:

```bash
amd.com/spx_nps1: 5
amd.com/cpx_nps4: 24
``` 
```

Author: Same answer as above

#### **Notes**

- If `resource_naming_strategy` is not passed using `DevicePluginArguments` field in CR, then device plugin will internally default to `single` resource naming strategy. This maintains backwards compatibility with earlier release of device plugin with reported resource name of `amd.com/gpu`
- If a node has GPUs which do not support partitioning, such as MI210, then the GPUs are reported under resource name `amd.com/gpu` regardless of the resource naming strategy

> - These different naming styles of resources, for example, `amd.com/cpx_nps1` should be followed when requesting for resources in a pod spec

Reviewer: ```suggestion
- These different naming styles of resources, for example, `amd.com/cpx_nps4` should be followed when requesting for resources in a pod spec
```
